README.md
[![github-ci][github-ci-img]][github-ci-url] [![NPM downloads][download-img]][download-url] [![NPM version][npm-version]][npm-version-url]

[github-ci-img]:https://img.shields.io/github/workflow/status/qiuysh/figlet-console-webpack-plugin/CI%20workflow?logo=github&style=flat
[github-ci-url]:https://img.shields.io/github/workflow/status/qiuysh/figlet-console-webpack-plugin/CI%20workflow
[download-img]: https://img.shields.io/npm/dm/figlet-console-webpack-plugin?style=flat
[download-url]: https://img.shields.io/npm/dm/figlet-console-webpack-plugin
[npm-version]: https://img.shields.io/npm/v/figlet-console-webpack-plugin?style=flat
[npm-version-url]: https://img.shields.io/npm/v/figlet-console-webpack-plugin


## figlet-console-webpack-plugin

### Introduction

A application version output in the browser console with figlet


### Install

Use npm or yarn install this webpack plugin.

```js
 npm install figlet-console-webpack-plugin --save-dev 
```


### Usage

Example

Set this in the webpack config file

```js

new FigletConsoleWebpackPlugin({
  name: 'console',
  content: 'version 0.0.1',
  options: {
    mark: '#'
  }
})

```

That should print out:

```
##################################################
                                        _        
   ___    ___    _ __    ___    ___   | |   ___ 
  / __|  / _ \  | '_ \  / __|  / _ \  | |  / _ \
 | (__  | (_) | | | | | \__ \ | (_) | | | |  __/
  \___|  \___/  |_| |_| |___/  \___/  |_|  \___|
                                                
##################################################
                  version 0.0.1                  
```


### Options


##### font 

You can set a string, default value: 'Standard'.
Only support: 'Ghost' and 'Standard'.


##### mark

You can set a string, default value: '#'.


#### color

You can set a string, default value: '#333'.

##### prodOnly

You can set a boolean, default value: false.
If you set true, that only print out in production environment.


##### markMaxLength

You can set a number, default value: 50.
This needs to be adjusted according to name with figlet.


CHANGELOG.md
# 更新历史 | changelog 


### [1.0.1](https://github.com/qiuysh/figlet-console-webpack-plugin/compare/v1.0.1-alpha...v1.0.1) (2022-10-26)


### ♻ Code Refactoring | 代码重构

* refactor output ([3f5ecc0](https://github.com/qiuysh/figlet-console-webpack-plugin/commit/3f5ecc00d96695b18813e37e4e53551db085b6f9))


### ✨ Features | 新功能

* add color config ([70aa165](https://github.com/qiuysh/figlet-console-webpack-plugin/commit/70aa165050f020f9728ec42d679e78293274df0f))
* add compilation assets index source type ([6bcf189](https://github.com/qiuysh/figlet-console-webpack-plugin/commit/6bcf189d97c08253f92491d25b5bfbff6a8fb2cd))
* remove try catch ([d18224e](https://github.com/qiuysh/figlet-console-webpack-plugin/commit/d18224e833cdbcd048cee4d0ea1d1505651f652a))
* rename figlet generate function name ([7b42b86](https://github.com/qiuysh/figlet-console-webpack-plugin/commit/7b42b861170e12131afc6882925a6d6f9d6a6eaa))

### [1.0.1-alpha](https://github.com/qiuysh/figlet-console-webpack-plugin/compare/v1.0.0-alpha...v1.0.1-alpha) (2022-04-09)


### ✨ Features | 新功能

* update type declaration ([7140479](https://github.com/qiuysh/figlet-console-webpack-plugin/commit/7140479e9689ac3f8f0c0117662ba36532d0eb9e))


.eslintrc.js
module.exports = {
  parser: "@typescript-eslint/parser",
  plugins: ["@typescript-eslint"],
  extends: [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:prettier/recommended",
  ],
  parserOptions: {
    project: "./tsconfig.json",
  },
  env: {
    es6: true,
    node: true,
    commonjs: true,
  },
  rules: {
    "prefer-const": "warn",
    "no-var": "warn",
    "one-var": [
      "warn",
      { var: "always", let: "never", const: "never" },
    ],
    "no-sequences": "warn",
    "no-cond-assign": "warn",
    "no-unused-vars": "warn",
    "no-unused-expressions": [
      "warn",
      { allowTernary: true },
    ],
    "no-useless-escape": "warn",
    "no-import-assign": "warn",
    "no-nested-ternary": "warn",
    "no-bitwise": "warn",
    "no-void": "warn",
    "no-restricted-globals": "warn",
    "yoda": "off",
    "prefer-destructuring": "warn",
    "class-methods-use-this": "warn",
  },
};


.versionrc.js
module.exports = {
  header: "# 更新历史 | changelog \n\n",
  types: [
    { type: "feat", section: "✨ Features | 新功能" },
    { type: "fix", section: "🐛 Bug Fixes | Bug 修复" },
    {
      type: "perf",
      section: "⚡ Performance Improvements | 性能优化",
      hidden: true,
    },
    {
      type: "revert",
      section: "⏪ Reverts | 回退",
      hidden: true,
    },
    {
      type: "chore",
      section: "📦 Chores | 其他更新",
      hidden: true,
    },
    {
      type: "docs",
      section: "📝 Documentation | 文档",
      hidden: true,
    },
    {
      type: "style",
      section: "💄 Styles | 风格",
      hidden: true,
    },
    {
      type: "refactor",
      section: "♻ Code Refactoring | 代码重构",
    },
    {
      type: "test",
      section: "✅ Tests | 测试",
      hidden: true,
    },
    {
      type: "build",
      section: "👷‍ Build System | 构建",
      hidden: true,
    },
    {
      type: "ci",
      section: "🔧 Continuous Integration | CI 配置",
      hidden: true,
    },
  ],
  skip: {
    // "bump": true, //缓存变化，并重置git状态至最近的tag节点
    commit: true, //提交变动
    tag: true, //在git中增加tag标识
  },
};


demo/index.js
const figlet = require("figlet");

figlet.text(
  "console",
  {
    horizontalLayout: "full",
    verticalLayout: "full",
  },
  (err, data) => {
    if (err) {
      console.log("Something went wrong...");
      console.dir(err);
      return;
    }
    let text = encodeURI(data);
    const content = "version 0.0.1";
    const markSpace = "#".repeat(50);
    const len =
      getCharCodeLength(markSpace) -
      getCharCodeLength(content);
    const singleLen = len / 2;
    const space = " ".repeat(singleLen);
    console.log(singleLen);
    console.log(markSpace);
    console.log(`%c ${decodeURI(text)}`, "color: #bada55;");
    console.log(markSpace);
    console.log(`${space}${content}${space}`);
    console.log("\n");
  }
);

// figlet.metadata('Standard', function (err, options, headerComment) {
//   if (err) {
//     console.log('something went wrong...');
//     console.dir(err);
//     return;
//   }
//   console.dir(options);
//   // console.log(headerComment);
// });

function getCharCodeLength(data) {
  var length = 0;
  Array.from(data).map(function (char) {
    if (char.charCodeAt(0) > 255) {
      //字符编码大于255，说明是双字节字符
      length += 2;
    } else {
      length++;
    }
  });

  return length;
}


src/index.ts
import figlet from "figlet";
import {
  FigletConsoleOptionsProps,
  FigletConsoleWebpackPluginProps,
} from "./typings";

const DEFAULT_OPTIONS: FigletConsoleOptionsProps = {
  font: "Standard",
  mark: "#",
  color: "#333",
  prodOnly: false,
  markMaxLength: 50,
};

class FigletConsoleWebpackPlugin {
  pluginName: string;
  name: string;
  content: string;
  options: FigletConsoleOptionsProps;
  constructor({
    name,
    content,
    options,
  }: FigletConsoleWebpackPluginProps) {
    this.pluginName = "FigletConsoleWebpackPlugin";
    this.name = name;
    // TODO remove content "
    this.content = content?.replace(/\"/g, "") || "";
    this.options = {
      ...DEFAULT_OPTIONS,
      ...options,
    };
  }

  apply = compiler => {
    const {
      options: { mode },
      hooks,
    } = compiler;
    const { prodOnly } = this.options;
    // development env and  prodOnly
    if (mode === "development" && prodOnly) {
      return;
    }
    hooks.emit.tapAsync(
      this.pluginName,
      (compilation, callback) => {
        const assetsIndexName = "index.html";
        const data: string = this.generateFiglet();
        const figletStr: string = this.formatPrint(data);

        let assetSource: string =
          compilation.assets[assetsIndexName].source();

        if (
          typeof assetSource === "object" &&
          Buffer.isBuffer(assetSource)
        ) {
          assetSource = Buffer.from(assetSource).toString();
        }

        compilation.assets[assetsIndexName] = {
          source: () => {
            const replaceMarkStr = "</html>";
            assetSource = assetSource?.replace(
              replaceMarkStr,
              ""
            );
            assetSource = assetSource
              .concat(figletStr)
              .concat(replaceMarkStr);

            return assetSource;
          },
          size: () => assetSource.length,
        };
        callback();
      }
    );
  };

  generateFiglet = () => {
    const defaultOption = {
      font: this.options?.font,
      horizontalLayout: "full",
      verticalLayout: "full",
    };
    const text = figlet.textSync(this.name, defaultOption);
    return encodeURIComponent(text);
  };

  formatPrint = data => {
    const { mark, markMaxLength, color } = this.options;
    const markSpace: string = mark?.repeat(markMaxLength);
    // Get mark width, we can set mark width equal figlet width
    const markSpaceLen: number =
      this.getCharCodeLength(markSpace);
    // content width
    const contentLen: number = this.getCharCodeLength(
      this.content
    );
    const halfRange: number =
      (markSpaceLen - contentLen) / 2;
    const spaceStr: string = " ".repeat(halfRange);

    let style;
    if (Array.isArray(color)) {
      style = [
        `background: linear-gradient(to right, ${color.join(",")})`,
        '-webkit-background-clip: text',
        'color:transparent',
      ].join(";")
    } else {
      style = [`color: ${color}`].join(";")
    }

    const outputStr = `
      <script>
        (function() {
          var text = decodeURIComponent("${data}");
          var styles = "${style}";
          var colorPrefix ="%c";
          console.log(colorPrefix + "${markSpace}", styles);
          console.log(colorPrefix + text, styles);
          console.log(colorPrefix + "${markSpace}", styles);
          console.log(colorPrefix + "${spaceStr}${this.content}${spaceStr}", styles);
          console.log("\\n");
        })()
      </script>
    `;
    
    return outputStr;
  };

  getCharCodeLength = data => {
    let length = 0;
    Array.from(data).map((char: any) => {
      if (char.charCodeAt(0) > 255) {
        length += 2;
      } else {
        length++;
      }
    });
    return length;
  };
}

export = FigletConsoleWebpackPlugin;
